<a href="https://colab.research.google.com/github/Cheetahzzzz1/Machine-Learning-Algo-on-Pizza.csv/blob/main/Machine_Learning_PizzaData.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Homework 3 Question 2-(a)

In [ ]:
import pandas as pd
file_path = "Pizza.csv"
pizza_data = pd.read_csv(file_path)

# We are displaying the first two rows to understand the columns
pizza_data.head()


brand     id   mois   prot    fat   ash  sodium  carb   cal
0     A  14069  27.82  21.43  44.87  5.11    1.77  0.77  4.93
1     A  14053  28.49  21.26  43.89  5.34    1.79  1.02  4.84
2     A  14025  28.35  19.99  45.78  5.08    1.63  0.80  4.95
3     A  14016  30.55  20.15  43.13  4.79    1.61  1.38  4.74
4     A  14005  30.49  21.28  41.65  4.82    1.64  1.76  4.67

In [ ]:
# Calculate the mean feature vector
mean_vector = pizza_data.iloc[:, 2:].mean()

mean_vector

mois      40.903067
prot      13.373567
fat       20.229533
ash        2.633233
sodium     0.669400
carb      22.864767
cal        3.271000
dtype: float64

The Mean Feature Vector m is given by:-



*   Moisture(moi)= 40.9030
*   Protein(prot)= 13.373567

*   Fat= 20.229533

*   Ash= 2.633233
*   Sodium= 0.6694

*   Carbohydrate(carb)= 22.864767
*   Calories(carb)= 3.271

In [ ]:
# Question 2-(b)

In [ ]:
import numpy as np

# Subtract the mean vector from each sample to center the data
data_centered = pizza_data.iloc[:, 2:] - mean_vector

# Calculate the scatter matrix (covariance matrix)
scatter_matrix = np.cov(data_centered.T)

# Calculate eigenvalues and eigenvectors of the scatter matrix
eigenvalues, eigenvectors = np.linalg.eig(scatter_matrix)

# Sort the eigenvalues and their corresponding eigenvectors in decreasing order
sorted_indices = np.argsort(eigenvalues)[::-1]
sorted_eigenvalues = eigenvalues[sorted_indices]
sorted_eigenvectors = eigenvectors[:, sorted_indices]

# Select the first two principal components (w1 and w2)
w1 = sorted_eigenvectors[:, 0]  # First principal component
w2 = sorted_eigenvectors[:, 1]  # Second principal component

w1, w2

(array([ 2.76963426e-01,  2.66941457e-01,  2.78933559e-01,  5.54340960e-02,
         1.11416057e-02, -8.78084364e-01,  6.03287596e-04]),
 array([ 0.74707368, -0.05573295, -0.65784531, -0.04060421, -0.02381376,
         0.00681755, -0.06125383]))

The first two principle components w1 and w2 are:-



*   w1 = [0.27696, 0.26694, 0.27893, 0.05543, 0.01114, -0.87808, 0.00060]
*   w2 = [0.74707, -0.05573, -0.65785, -0.04060, -0.02381, 0.00682, -0.06125]

In [ ]:
# Question 2-(c)

In [ ]:
# Calculate coefficients a1 and a2 for each sample
a1 = np.dot(data_centered, w1)
a2 = np.dot(data_centered, w2)

# Reconstruct the feature vector for each sample using the first two principal components
approximations = mean_vector.values + np.outer(a1, w1) + np.outer(a2, w2)

# Compare the original and approximated feature vectors for the first two samples
original_samples = pizza_data.iloc[:2, 2:]
approx_samples = pd.DataFrame(approximations[:2], columns=original_samples.columns)

# Display original and approximated values for the first two samples
comparison = pd.concat([original_samples.reset_index(drop=True), approx_samples.reset_index(drop=True)], axis=1)
comparison.columns = ['Original_Mois', 'Original_Prot', 'Original_Fat', 'Original_Ash', 'Original_Sodium', 'Original_Carb', 'Original_Cal',
                      'Approx_Mois', 'Approx_Prot', 'Approx_Fat', 'Approx_Ash', 'Approx_Sodium', 'Approx_Carb', 'Approx_Cal']

print("Original vs Approximated Samples:")
print(comparison)

Original vs Approximated Samples:
   Original_Mois  Original_Prot  Original_Fat  Original_Ash  Original_Sodium  \
0          27.82          21.43         44.87          5.11             1.77   
1          28.49          21.26         43.89          5.34             1.79   

   Original_Carb  Original_Cal  Approx_Mois  Approx_Prot  Approx_Fat  \
0           0.77          4.93    27.783497    21.528516   44.827343   
1           1.02          4.84    28.550099    21.373606   43.974663   

   Approx_Ash  Approx_Sodium  Approx_Carb  Approx_Cal  
0    5.105078       1.585889     0.772237    4.928370  
1    5.039462       1.554668     1.078440    4.857595  


In [ ]:
# Standardize the data: subtract mean and divide by standard deviation for each feature
data_standardized = (pizza_data.iloc[:, 2:] - mean_vector) / pizza_data.iloc[:, 2:].std()

# Calculate the scatter matrix for the standardized data
scatter_matrix_standardized = np.cov(data_standardized.T)

# Calculate eigenvalues and eigenvectors of the standardized scatter matrix
eigenvalues_std, eigenvectors_std = np.linalg.eig(scatter_matrix_standardized)

# Sort eigenvalues and their corresponding eigenvectors in decreasing order
sorted_indices_std = np.argsort(eigenvalues_std)[::-1]
sorted_eigenvalues_std = eigenvalues_std[sorted_indices_std]
sorted_eigenvectors_std = eigenvectors_std[:, sorted_indices_std]

# Select the first two principal components for the standardized data
w1_std = sorted_eigenvectors_std[:, 0]  # First principal component
w2_std = sorted_eigenvectors_std[:, 1]  # Second principal component

# Calculate coefficients a1 and a2 for each standardized sample
a1_std = np.dot(data_standardized, w1_std)
a2_std = np.dot(data_standardized, w2_std)

# Reconstruct the feature vector for each standardized sample using the first two principal components
approximations_std = mean_vector.values + np.outer(a1_std, w1_std) * pizza_data.iloc[:, 2:].std().values + np.outer(a2_std, w2_std) * pizza_data.iloc[:, 2:].std().values

# Compare original and approximated feature vectors for the first two samples after standardization
approx_samples_std = pd.DataFrame(approximations_std[:2], columns=original_samples.columns)
comparison_std = pd.concat([original_samples.reset_index(drop=True), approx_samples_std.reset_index(drop=True)], axis=1)
comparison_std.columns = ['Original_Mois', 'Original_Prot', 'Original_Fat', 'Original_Ash', 'Original_Sodium', 'Original_Carb', 'Original_Cal',
                          'Approx_Std_Mois', 'Approx_Std_Prot', 'Approx_Std_Fat', 'Approx_Std_Ash', 'Approx_Std_Sodium', 'Approx_Std_Carb', 'Approx_Std_Cal']

print("Original vs Approximated Samples:")
print(comparison)

Original vs Approximated Samples:
   Original_Mois  Original_Prot  Original_Fat  Original_Ash  Original_Sodium  \
0          27.82          21.43         44.87          5.11             1.77   
1          28.49          21.26         43.89          5.34             1.79   

   Original_Carb  Original_Cal  Approx_Mois  Approx_Prot  Approx_Fat  \
0           0.77          4.93    27.783497    21.528516   44.827343   
1           1.02          4.84    28.550099    21.373606   43.974663   

   Approx_Ash  Approx_Sodium  Approx_Carb  Approx_Cal  
0    5.105078       1.585889     0.772237    4.928370  
1    5.039462       1.554668     1.078440    4.857595  


In [ ]:
# Question 2-(d)

In [ ]:
# Standardize the data: subtract mean and divide by standard deviation for each feature
data_standardized = (pizza_data.iloc[:, 2:] - mean_vector) / pizza_data.iloc[:, 2:].std()

# Calculate the scatter matrix for the standardized data
scatter_matrix_standardized = np.cov(data_standardized.T)

# Calculate eigenvalues and eigenvectors of the standardized scatter matrix
eigenvalues_std, eigenvectors_std = np.linalg.eig(scatter_matrix_standardized)

# Sort eigenvalues and their corresponding eigenvectors in decreasing order
sorted_indices_std = np.argsort(eigenvalues_std)[::-1]
sorted_eigenvalues_std = eigenvalues_std[sorted_indices_std]
sorted_eigenvectors_std = eigenvectors_std[:, sorted_indices_std]

# Select the first two principal components for the standardized data
w1_std = sorted_eigenvectors_std[:, 0]  # First principal component
w2_std = sorted_eigenvectors_std[:, 1]  # Second principal component

# Calculate coefficients a1 and a2 for each standardized sample
a1_std = np.dot(data_standardized, w1_std)
a2_std = np.dot(data_standardized, w2_std)

# Reconstruct the feature vector for each standardized sample using the first two principal components
approximations_std = mean_vector.values + np.outer(a1_std, w1_std) * pizza_data.iloc[:, 2:].std().values + np.outer(a2_std, w2_std) * pizza_data.iloc[:, 2:].std().values

# Compare original and approximated feature vectors for the first two samples after standardization
approx_samples_std = pd.DataFrame(approximations_std[:2], columns=original_samples.columns)
comparison_std = pd.concat([original_samples.reset_index(drop=True), approx_samples_std.reset_index(drop=True)], axis=1)
comparison_std.columns = ['Original_Mois', 'Original_Prot', 'Original_Fat', 'Original_Ash', 'Original_Sodium', 'Original_Carb', 'Original_Cal',
                          'Approx_Std_Mois', 'Approx_Std_Prot', 'Approx_Std_Fat', 'Approx_Std_Ash', 'Approx_Std_Sodium', 'Approx_Std_Carb', 'Approx_Std_Cal']

print("Original vs Approximated Samples:")
print(comparison)

Original vs Approximated Samples:
   Original_Mois  Original_Prot  Original_Fat  Original_Ash  Original_Sodium  \
0          27.82          21.43         44.87          5.11             1.77   
1          28.49          21.26         43.89          5.34             1.79   

   Original_Carb  Original_Cal  Approx_Mois  Approx_Prot  Approx_Fat  \
0           0.77          4.93    27.783497    21.528516   44.827343   
1           1.02          4.84    28.550099    21.373606   43.974663   

   Approx_Ash  Approx_Sodium  Approx_Carb  Approx_Cal  
0    5.105078       1.585889     0.772237    4.928370  
1    5.039462       1.554668     1.078440    4.857595  
